### PR TITLE
Basic C++11 features

### PR DESCRIPTION
**Pull request by rleigh-codelibre**
Next part of [Minimal C++11 features](https://trello.com/c/h4pHPV2h/42-minimal-c-11-features).  This PR switches to using a few simple C++11 features.

- Use `<array>`, `<cstdint>`, `<memory>` and `<tuple>` in place of `<ome/compat/array.h>`, `<ome/compat/cstdint.h>`, `<ome/compat/memory.h>` and `<ome/compat/tuple.h>`, respectively (with namespace change from `ome::compat` to `std` for the corresponding types)
- Use `>>` in place of `> >` in templates (spaces previously required)
- Use initialiser lists when constructing objects, mainly containers.  This uses `foo{elem1, elem2, elemn}` in place of `foo` to directly fill containers with elements.  Mainly used with parameterised test data which had to use static arrays as the data source.
- Use range-based for loop and `auto` type rather than manually define complex iterator types and manually iterate; this greatly simplifies `for` loops
- Use `enum class` to scope enum values; previously they leaked into the enclosing scope
- Deprecate compatibility headers (in doxygen comments)

The compatibility headers were removed to test that all downstream `basic-c++11` branches in ome-model, ome-files and ome-qtwidgets components were using only the standard headers, then added back.  The downstream components continue to build with this PR alone, demonstrating that the compatibility headers continue to function, and that nothing breaks as a result of the changes here.

This will be followed by opening PRs against the downstream components in sequence.  Some additional features such as `<thread>` usage will be done in later PRs.

--------

Testing: Check all builds are green.

### CODE REVIEW

File: lib/ome/common/xml/dom/Base.h [hunk 1]
```diff
@@ -151,7 +151,29 @@ namespace ome
...
           operator bool () const
           {
-            return get() != 0;
+            return get() != nullptr;
+          }
+
+          /**
+           * Check if the wrapped type is NULL.
+           *
+           * @returns true if valid, false if NULL.
```
Comment (mtbc) [inline lib/ome/common/xml/dom/Base.h]: Is this the correct way around?

Comment (rleigh-codelibre) [reply lib/ome/common/xml/dom/Base.h]: No, it's a copy paste error, and also confusing.  I've pushed a commit to correct and reword it.

**Comment by mtbc** [comment]
Good to merge.

**Comment by rleigh-codelibre** [comment]
@sbesson Is anything blocking this being merged, or can we go ahead with the merge?

**Comment by sbesson** [comment]
Overall looks good and certainly good to reduce the usage of our own `ome::compat` in favor of the C++ standard library. As some of these changes have been indirectly impacting the process of [inclusion of ome-files-py into the superbuild](https://github.com/ome/ome-cmake-superbuild/pull/115), I will leave this PR opened until we sort out the compiler option.

**Comment by rleigh-codelibre** [comment]
Note that the ome-files-py problem will exist while the PR is open or merged; it makes no difference.

**Pull request merged**